service/raft/raft_group0_client.hh

```diff
@@ -190,4 +191,54 @@ public:
     void set_query_result(utils::UUID query_id, service::broadcast_tables::query_result qr);
 };
 
+using mutations_generator = coroutine::experimental::generator<mutation>;
+
+// mutations_collector is used to gather mutations which are side effects
+// of functions execution. They need to be announced under single guard
+// for atomicity. As functions which produce mutations may embed each other
+// we need to decouple announcing step to a common external place here.
+// It also supports generator callbacks to avoid holding too many mutations
+// in memory.
+//
+// Single mutations_collector object represents a single transaction.
+// If size or number of mutations is too big for raft too handle it will be
+// rejected.
+class mutations_collector {
+public:
+    using generator_func = std::function<mutations_generator(api::timestamp_type t)>;
+private:
+    std::vector<mutation> _muts;
+    std::vector<generator_func> _generators;
+    std::optional<::service::group0_guard> _guard;
+
+    future<> materialize_mutations();
+public:
+    mutations_collector(::service::group0_guard&& g) : _guard(std::move(g)) {}
+    // Constructor with optional guard used to handle both legacy and current code.
+    // There is no guard for legacy code but the whole class may be passed
+    // through to simplify the flow.
+    mutations_collector(std::optional<::service::group0_guard> g) : _guard(std::move(g)) {}
```

Comment:
Single-argument constructors should be `explicit` to prevent accidental conversions.

Reply:
ok